### ISSUE TITLE
Azure Pipelines Azure CLI task - no colors when doing Bicep deployment

### ISSUE THREAD

**Issue by chudytom** [microsoft/azure-pipelines-tasks#15473]
## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: AzureCLI@2

## Environment
- Server - Azure Pipelines
    
    - If using Azure Pipelines, provide the account name, team project name, build definition name/build number: 
    dongenergy-p/Power Hub/Pipelines/flexp-infra-bicep/20211103.7

- Agent - Hosted or Private: 
    
    - If using Hosted agent, provide agent queue name: MS Hosted agents - Ubuntu-20.04, from Azure Pipelines agent pool

## Issue Description
I tried to deployed infrastructure using Azure Bicep. I did it using Azure CLI and specifically AzureCLI@2 task. It works, however when I use --what-if functionality the changes I see on the screen have no colors, which make the output much harder to read. When I run the same commands from my local CLI all the colors are visible.

Command used - 
`az deployment group create -f infrastructure/bicep/<project-name>.bicep --subscription=<project-name> -g <project-name> --parameters env=dev --mode=complete --name=<project-name>_20211103_1243 --what-if
`

Current behavior:
![image](https://user-images.githubusercontent.com/29259905/140061435-1b996336-a4ef-4ba0-91ac-b68333015d6f.png)

Expected behavior (from my local machine)
![image](https://user-images.githubusercontent.com/29259905/140061484-e1592d47-fa5e-4240-8c8e-c42c9d417bda.png)

### Task logs

Will provide it later if really needed

## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

Done

### Error logs

No errors even after enabling system diagnostics

**Comment by brettpostin** [comment]
This isn't specific to bicep templates either. I have a release stage to do a what-if analysis for approval on a compiled bicep template (so json).

The colour coding would be really helpful.

**Comment by chudytom** [comment]
@bishal-pdMSFT is there maybe any action planned on this issue?

**Comment by bishal-pdMSFT** [comment]
@naga-bandi can you please ask someone to look at thia

**Comment by NAGA-BANDI** [comment]
Ack'ed @***@***.******@***.***>.

@Renukadevi Vairavelu (MINDTREE ***@***.***>: Please look into this.

Thanks,
Naga Bandi | +91 99 89 233 944

**Comment by NAGA-BANDI** [comment]
Sure Naga, will check on this

Regards,
Renukadevi V

From: Naga Varaprasad Bandi (MINDTREE LIMITED) ***@***.***>
Sent: Sunday, August 7, 2022 11:40 PM
To: microsoft/azure-pipelines-tasks ***@***.***>; microsoft/azure-pipelines-tasks ***@***.***>; ***@***.***; Renukadevi Vairavelu (MINDTREE LIMITED) ***@***.***>
Cc: Mention ***@***.***>
Subject: RE: [microsoft/azure-pipelines-tasks] Azure Pipelines Azure CLI task - no colors when doing Bicep deployment (Issue #15473)

Ack'ed @***@***.******@***.***>.

@Renukadevi Vairavelu (MINDTREE ***@***.***>: Please look into this.

Thanks,
Naga Bandi | +91 99 89 233 944

**Comment by htuomola** [comment]
@v-nagarajku Colors are also not showing on hosted ubuntu-latest agent either, with task AzureCLI@2 

```yaml 
- task: AzureCLI@2
  inputs:
    azureSubscription: <connection name>
    scriptType: bash
    scriptLocation: inlineScript
    inlineScript: |
      az --version
      az deployment sub create --what-if --location <....params>
```

![image](https://user-images.githubusercontent.com/1266311/190647398-6c995a46-1e81-41ea-b8fe-13aea13cd346.png)

also running on hosted windows-latest agent with `scriptType: pscore` makes no difference.

vs. in windows local terminal (ps core)
![image](https://user-images.githubusercontent.com/1266311/190647308-2fc10df5-585a-4599-a401-1c44a76e539e.png)

**Comment by ekersale** [comment]
Any update here ?

**Comment by NAGA-BANDI** [comment]
@***@***.***>.

Thanks,
Naga Bandi | +91 99 89 233 944
***@***.***

**Comment by rvairavelu** [comment]
@ekersale. This feature is not implemented.  At present, we could not prioritize it, due to other critical issues in our queue. 
cc  @vijayma @NAGA-BANDI @v-nagarajku

**Comment by Kravca** [comment]
Anybody got some workaround for this?

**Comment by roofiq** [comment]
Up, it would be really handy

**Comment by Rod-Sychev** [comment]
Come on MSFT, this is basics!

**Comment by miniGweek** [comment]
MSFT Team, this is important!
While running the What-if step of bicep templates in our pipeline we need to see a colored output to make sense of what's getting modified, deleted or added! This is basics!
Please prioritize this!!

**Comment by arwinters** [comment]
Yes MSFT we are waiting on it.

**Comment by TiTi** [comment]
Any update on this topic ?

Azure declarative IaC language which is recommended is Bicep (see green box here: https://learn.microsoft.com/en-us/azure/templates/)
Unfortunately no color in Azure DevOps :-/
This is a real turn down for team work with centralized pipeline!

Can we please have a little bit of color for bicep in Azure DevOps ?

Note: color is working with terraform....

**Comment by anthony-c-martin** [comment]
I did some digging on this, and chatted to @JamieMagee who gave me some great pointers.

Current theory:
- The AzureCLI task (as well as other MS-published tasks) do use Node's `child_process` directly - instead they use a wrapper library to interact with the shell ([see here](https://github.com/microsoft/azure-pipelines-tasks/blob/3bea84b6178187eeae55ccaf4dc989f7accac5ef/Tasks/AzureCLIV2/src/ScriptType.ts#L2))
- The wrapper library is invoked [here](https://github.com/microsoft/azure-pipelines-tasks/blob/3bea84b6178187eeae55ccaf4dc989f7accac5ef/Tasks/AzureCLIV2/azureclitask.ts#L48-L55)
- Here's where the wrapper library uses `child_process` in the `exec()` method: [see here](https://github.com/microsoft/azure-pipelines-task-lib/blob/f1a07a7ef97d603242ca4c877bf5987acd923ef3/node/toolrunner.ts#L1218)
- Theory is that the indirection or capturing of stdout/stderr here impacts Node's built-in mechanism to automatically detect shell color support.

We do not believe this was a deliberate decision - just an accidental result of the indirection.

An interesting test would be to try and set the env variable `FORCE_COLOR` to `1` to override Node's built-in color detection: https://nodejs.org/api/tty.html#writestreamgetcolordepthenv.

**Comment by TiTi** [comment]
Thanks for digging this @anthony-c-martin . Just tried it without success :/

**Comment by slavizh** [comment]
There is some not pretty workaround for PowerShell https://paul-mackinnon.medium.com/azure-devops-ansi-colour-coding-df8ef7406422
It is not pretty as not everything is 100% as would appear in regular PowerShell console. Overall the right way would be ADO to support 8bit color codes instead of Bicep team or we implement workarounds.

**Comment by TiTi** [comment]
Whatever I try, I can't output the escape sequence (ex: ``e[34m) even when trying to save **az cli** output in a file. (`az deployment group what-if ...`)
Thus the powershell workaround doesn't even work for az cli.
(i couldn't use New-Az[Subscription]Deployment because it didn't handle .bicepparam until recently ; and I just tried but got other issues preventing me to use it, it's a mess!)

It would be nice to have an option like PY_COLORS or ANSIBLE_FORCE_COLOR  
"This option forces color mode even when running without a TTY"

**Comment by vuylstekeb** [comment]
+1 would be really useful to have

**Comment by Saulopv** [comment]
+1 basic feature really..

**Comment by SaraGonc** [comment]
+ This will be very helpful. Please MSFT! 🥇

**Comment by 0-0-1-0-1-0-1-0** [comment]
**NOTE: I haven't solved an - on resources, if someone can get me a log sample I think i can craft a regex to account for - <resource> <api version> lines, for now i only cover ~ + and `*, this is an improvement over the article linked below.**

For those that want some color in your life:

This is the task I use.

Tested on a self-hosted windows Agent.

PowerShell 7.5 is installed on the agent and so is the Az module, using a main.bicepparam file:

**This works using PWSH (7+) or Regular Powershell (5.1)**

Simply change these lines:

**Powershell7+**
`      pwsh: true`

`        $output = pwsh -Command {New-AzResourceGroupDeployment -DeploymentDebugLogLevel All -ResourceGroupName "${{ parameters.component.resourceGroupName }}" -TemplateParameterFile "${{ parameters.component.parametersFile }}" -Name '$(Build.SourceVersion)-$(build.BuildId)' -WhatIf}`

to

**Powershell 5.1**
`      pwsh: false`

`        $output = powershell -Command {New-AzResourceGroupDeployment -DeploymentDebugLogLevel All -ResourceGroupName "${{ parameters.component.resourceGroupName }}" -TemplateParameterFile "${{ parameters.component.parametersFile }}" -Name '$(Build.SourceVersion)-$(build.BuildId)' -WhatIf}`

___

_**NOTE: ${{ parameters.component.X }} will not exist in your environment sub that out for however you set parameters/variables for your tasks.**_

___

This task will download the latest Bicep CLI exe and set the path to it, our build agents run under a service account so the $env:USERPROFILE\.bicep works but you can set the path to anything you want. If you globally install the bicep on the build agents , you can remove that section of the script entirely.

Credit: https://paul-mackinnon.medium.com/azure-devops-ansi-colour-coding-df8ef7406422

```
  - task: AzurePowerShell@5
    displayName: 'Bicep What-If Analysis'
    inputs:
      azureSubscription: ${{ parameters.component.azureSubscription }}
      scriptType: 'InlineScript'
      azurePowerShellVersion: LatestVersion
      pwsh: true
      Inline: | # powershell
        $installPath = "$env:USERPROFILE\.bicep"
        Write-Host "Installing Bicep CLI to '$installPath'"

        if (-not (Test-Path -Path $installPath)) {
          Write-Host "Creating install directory ..."
          $installDir = New-Item -ItemType Directory -Path $installPath -Force
          $installDir.Attributes += 'Hidden'
        }

        $latestVersion = (Invoke-RestMethod -Uri "https://api.github.com/repos/Azure/bicep/releases/latest").tag_name
        Write-Host "Downloading Bicep CLI with latest version: $latestVersion ..."
        (New-Object Net.WebClient).DownloadFile("https://github.com/Azure/bicep/releases/latest/download/bicep-win-x64.exe", "$installPath\bicep.exe")

        Write-Host "Adding Bicep CLI to PATH ..."
        $currentPath = (Get-Item -path "HKCU:\Environment" ).GetValue('Path', '', 'DoNotExpandEnvironmentNames')
        if (-not $currentPath.Contains("%USERPROFILE%\.bicep")) { setx PATH ($currentPath + ";%USERPROFILE%\.bicep") }
        if (-not $env:path.Contains($installPath)) { $env:path += ";$installPath" }

        bicep --version
        
        Write-Host "Using Bicep File: '${{ parameters.component.parametersFile }}'"
        $output = pwsh -Command {New-AzResourceGroupDeployment -DeploymentDebugLogLevel All -ResourceGroupName "${{ parameters.component.resourceGroupName }}" -TemplateParameterFile "${{ parameters.component.parametersFile }}" -Name '$(Build.SourceVersion)-$(build.BuildId)' -WhatIf}

        $output `
        -replace "\[38;5;208m", "[31;1m" `
        -replace "\[38;5;77m" , "[32;1m" `
        -replace "\[38;5;141m", "[33;1m" `
        -replace "\[38;5;246m", "[34;1m" `
        -replace "\[38;5;136m", "[35;1m" `
        -replace "(?m)(^\s*~.*$([char]0x1b)\[33;1m$([char]0x1b)\[0m)", "$([char]0x1b)[33;1m$&" `
        -replace "(?m)^(  *\+.*$([char]0x1b)\[32;1m)", "$([char]0x1b)[32;1m$&" `
        -replace "(?m)^\s*\*.*?(?=\n)", "$([char]0x1b)[34;1m$&"
```

**Comment by 0-0-1-0-1-0-1-0** [comment]
I figured I would throw this in here as well. I went and modified this module code (the posh script) https://github.com/EvotecIT/PSWriteColor

To add support for ANSI 4BIT color codes. You then could use this module in your PowerShell scripts to give meaningful color coding to your output.

Create a Write-Color.psm1 file and paste the below code into. Save the file into the system wide modules folder so it's loaded at PoSH start.

This is assuming you are using a self-hosted build agent, if not you will need to figure out how to load this into the azure hosted agents.

Then you can simply use Write-Color with the -ANSI switch instead of Write-Host for any color specific output.

I let the author of the original module know about this, hopefully he merges it into his code, and you can simply Install-Module PSWriteColor on the azure hosted agents.

```
function Write-Color {
    <#
    .SYNOPSIS
    Write-Color is a wrapper around Write-Host delivering a lot of additional features for easier color options.

    .DESCRIPTION
    Write-Color is a wrapper around Write-Host delivering a lot of additional features for easier color options.

    It provides:
    - Easy manipulation of colors,
    - Logging output to file (log)
    - Nice formatting options out of the box.
    - Ability to use aliases for parameters

    .PARAMETER Text
    Text to display on screen and write to log file if specified.
    Accepts an array of strings.

    .PARAMETER Color
    Color of the text. Accepts an array of colors. If more than one color is specified it will loop through colors for each string.
    If there are more strings than colors it will start from the beginning.
    Available colors are: Black, DarkBlue, DarkGreen, DarkCyan, DarkRed, DarkMagenta, DarkYellow, Gray, DarkGray, Blue, Green, Cyan, Red, Magenta, Yellow, White

    .PARAMETER BackGroundColor
    Color of the background. Accepts an array of colors. If more than one color is specified it will loop through colors for each string.
    If there are more strings than colors it will start from the beginning.
    Available colors are: Black, DarkBlue, DarkGreen, DarkCyan, DarkRed, DarkMagenta, DarkYellow, Gray, DarkGray, Blue, Green, Cyan, Red, Magenta, Yellow, White

    .PARAMETER StartTab
    Number of tabs to add before text. Default is 0.

    .PARAMETER LinesBefore
    Number of empty lines before text. Default is 0.

    .PARAMETER LinesAfter
    Number of empty lines after text. Default is 0.

    .PARAMETER StartSpaces
    Number of spaces to add before text. Default is 0.

    .PARAMETER LogFile
    Path to log file. If not specified no log file will be created.

    .PARAMETER DateTimeFormat
    Custom date and time format string. Default is yyyy-MM-dd HH:mm:ss

    .PARAMETER LogTime
    If set to $true it will add time to log file. Default is $true.

    .PARAMETER LogRetry
    Number of retries to write to log file, in case it can't write to it for some reason, before skipping. Default is 2.

    .PARAMETER Encoding
    Encoding of the log file. Default is Unicode.

    .PARAMETER ShowTime
    Switch to add time to console output. Default is not set.

    .PARAMETER NoNewLine
    Switch to not add new line at the end of the output. Default is not set.

    .PARAMETER NoConsoleOutput
    Switch to not output to console. Default all output goes to console.

    .EXAMPLE
    Write-Color -Text "Red ", "Green ", "Yellow " -Color Red,Green,Yellow

    .EXAMPLE
    Write-Color -Text "This is text in Green ",
                      "followed by red ",
                      "and then we have Magenta... ",
                      "isn't it fun? ",
                      "Here goes DarkCyan" -Color Green,Red,Magenta,White,DarkCyan

    .EXAMPLE
    Write-Color -Text "This is text in Green ",
                      "followed by red ",
                      "and then we have Magenta... ",
                      "isn't it fun? ",
                      "Here goes DarkCyan" -Color Green,Red,Magenta,White,DarkCyan -StartTab 3 -LinesBefore 1 -LinesAfter 1

    .EXAMPLE
    Write-Color "1. ", "Option 1" -Color Yellow, Green
    Write-Color "2. ", "Option 2" -Color Yellow, Green
    Write-Color "3. ", "Option 3" -Color Yellow, Green
    Write-Color "4. ", "Option 4" -Color Yellow, Green
    Write-Color "9. ", "Press 9 to exit" -Color Yellow, Gray -LinesBefore 1

    .EXAMPLE
    Write-Color -LinesBefore 2 -Text "This little ","message is ", "written to log ", "file as well." `
                -Color Yellow, White, Green, Red, Red -LogFile "C:\testing.txt" -TimeFormat "yyyy-MM-dd HH:mm:ss"
    Write-Color -Text "This can get ","handy if ", "want to display things, and log actions to file ", "at the same time." `
                -Color Yellow, White, Green, Red, Red -LogFile "C:\testing.txt"

    .EXAMPLE
    Write-Color -T "My text", " is ", "all colorful" -C Yellow, Red, Green -B Green, Green, Yellow
    Write-Color -t "my text" -c yellow -b green
    Write-Color -text "my text" -c red

    .EXAMPLE
    Write-Color -Text "Testuję czy się ładnie zapisze, czy będą problemy" -Encoding unicode -LogFile 'C:\temp\testinggg.txt' -Color Red -NoConsoleOutput

    .NOTES
    Understanding Custom date and time format strings: https://learn.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings
    Project support: https://github.com/EvotecIT/PSWriteColor
    Original idea: Josh (https://stackoverflow.com/users/81769/josh)

    #>
    [alias('Write-Colour')]
    [CmdletBinding()]
    param (
        [alias ('T')] [String[]]$Text,
        [alias ('C', 'ForegroundColor', 'FGC')] [ConsoleColor[]]$Color = [ConsoleColor]::White,
        [alias ('B', 'BGC')] [ConsoleColor[]]$BackGroundColor = $null,
        [alias ('Indent')][int] $StartTab = 0,
        [int] $LinesBefore = 0,
        [int] $LinesAfter = 0,
        [int] $StartSpaces = 0,
        [alias ('L')] [string] $LogFile = '',
        [Alias('DateFormat', 'TimeFormat')][string] $DateTimeFormat = 'yyyy-MM-dd HH:mm:ss',
        [alias ('LogTimeStamp')][bool] $LogTime = $true,
        [int] $LogRetry = 2,
        [ValidateSet('unknown', 'string', 'unicode', 'bigendianunicode', 'utf8', 'utf7', 'utf32', 'ascii', 'default', 'oem')][string]$Encoding = 'Unicode',
        [switch] $ShowTime,
        [switch] $NoNewLine,
        [alias('HideConsole')][switch] $NoConsoleOutput,
        [switch] $ANSI
    )
    if (-not $NoConsoleOutput) {
        $DefaultColor = $Color[0]
        if ($null -ne $BackGroundColor -and $BackGroundColor.Count -ne $Color.Count) {
            Write-Error "Colors, BackGroundColors parameters count doesn't match. Terminated."
            return
        }
        if ($LinesBefore -ne 0) { for ($i = 0; $i -lt $LinesBefore; $i++) { Write-Host -Object "`n" -NoNewline } } # Add empty line before
        if ($StartTab -ne 0) { for ($i = 0; $i -lt $StartTab; $i++) { Write-Host -Object "`t" -NoNewline } }  # Add TABS before text
        if ($StartSpaces -ne 0) { for ($i = 0; $i -lt $StartSpaces; $i++) { Write-Host -Object ' ' -NoNewline } }  # Add SPACES before text
        if ($ShowTime) { Write-Host -Object "[$([datetime]::Now.ToString($DateTimeFormat))] " -NoNewline } # Add Time before output

        if ($ANSI) {
            $ANSIColors = @{
                Black = "`e[30m"; DarkBlue = "`e[34m"; DarkGreen = "`e[32m"; DarkCyan = "`e[36m";
                DarkRed = "`e[31m"; DarkMagenta = "`e[35m"; DarkYellow = "`e[33m"; Gray = "`e[37m";
                DarkGray = "`e[90m"; Blue = "`e[94m"; Green = "`e[92m"; Cyan = "`e[96m";
                Red = "`e[91m"; Magenta = "`e[95m"; Yellow = "`e[93m"; White = "`e[97m"
            }
            $ANSIBackgroundColors = @{
                Black = "`e[40m"; DarkBlue = "`e[44m"; DarkGreen = "`e[42m"; DarkCyan = "`e[46m";
                DarkRed = "`e[41m"; DarkMagenta = "`e[45m"; DarkYellow = "`e[43m"; Gray = "`e[47m";
                DarkGray = "`e[100m"; Blue = "`e[104m"; Green = "`e[102m"; Cyan = "`e[106m";
                Red = "`e[101m"; Magenta = "`e[105m"; Yellow = "`e[103m"; White = "`e[107m"
            }
            $ANSIReset = "`e[0m"
            for ($i = 0; $i -lt $Text.Length; $i++) {
                $colorCode = $ANSIColors[$Color[$i].ToString()]
                $bgColorCode = if ($BackGroundColor) { $ANSIBackgroundColors[$BackGroundColor[$i].ToString()] } else { "" }
                [string]$newString = $Text[$i].Replace("`n", "`n${colorCode}${bgColorCode}")
                $outputText = "${colorCode}${bgColorCode}${newString}${ANSIReset}"
                Write-Host -Object $outputText -NoNewline
            }
        } else {
            if ($Text.Count -ne 0) {
                if ($Color.Count -ge $Text.Count) {
                    # the real deal coloring
                    if ($null -eq $BackGroundColor) {
                        for ($i = 0; $i -lt $Text.Length; $i++) {
                            Write-Host -Object $Text[$i] -ForegroundColor $Color[$i] -NoNewline
                        }
                    } else {
                        for ($i = 0; $i -lt $Text.Length; $i++) {
                            Write-Host -Object $Text[$i] -ForegroundColor $Color[$i] -BackgroundColor $BackGroundColor[$i] -NoNewline
                        }
                    }
                } else {
                    if ($null -eq $BackGroundColor) {
                        for ($i = 0; $i -lt $Color.Length ; $i++) {
                            Write-Host -Object $Text[$i] -ForegroundColor $Color[$i] -NoNewline
                        }
                        for ($i = $Color.Length; $i -lt $Text.Length; $i++) {
                            Write-Host -Object $Text[$i] -ForegroundColor $DefaultColor -NoNewline
                        }
                    } else {
                        for ($i = 0; $i -lt $Color.Length ; $i++) {
                            Write-Host -Object $Text[$i] -ForegroundColor $Color[$i] -BackgroundColor $BackGroundColor[$i] -NoNewline
                        }
                        for ($i = $Color.Length; $i -lt $Text.Length; $i++) {
                            Write-Host -Object $Text[$i] -ForegroundColor $DefaultColor -BackgroundColor $BackGroundColor[0] -NoNewline
                        }
                    }
                }
            }
        }
        if ($NoNewLine -eq $true) {
            Write-Host -NoNewline
        } else {
            Write-Host
        }
        if ($LinesAfter -ne 0) { for ($i = 0; $i -lt $LinesAfter; $i++) { Write-Host -Object "`n" -NoNewline } }  # Add empty line after
    }
    if ($Text.Count -and $LogFile) {
        # Save to file
        $TextToFile = ""
        for ($i = 0; $i -lt $Text.Length; $i++) {
            $TextToFile += $Text[$i]
        }
        $Saved = $false
        $Retry = 0
        Do {
            $Retry++
            try {
                if ($LogTime) {
                    "[$([datetime]::Now.ToString($DateTimeFormat))] $TextToFile" | Out-File -FilePath $LogFile -Encoding $Encoding -Append -ErrorAction Stop -WhatIf:$false
                } else {
                    "$TextToFile" | Out-File -FilePath $LogFile -Encoding $Encoding -Append -ErrorAction Stop -WhatIf:$false
                }
                $Saved = $true
            } catch {
                if ($Saved -eq $false -and $Retry -eq $LogRetry) {
                    Write-Warning "Write-Color - Couldn't write to log file $($_.Exception.Message). Tried ($Retry/$LogRetry))"
                } else {
                    Write-Warning "Write-Color - Couldn't write to log file $($_.Exception.Message). Retrying... ($Retry/$LogRetry)"
                }
            }
        } Until ($Saved -eq $true -or $Retry -ge $LogRetry)
    }
}

Export-ModuleMember -Function '*' -Alias '*'#Get public and private function definition files.
```

**Comment by tknerr** [comment]
@TiTi reading [your comment above](https://github.com/microsoft/azure-pipelines-tasks/issues/15473#issuecomment-1970789538), do I read it correctly that you managed to force color output (even though not properly escaped) with azure cli tools (i.e. `az deployment sub create ... --what-if`)?

Was wondering if colors could be forced with azure cli based deployment, but didn't get so far yet...

**Comment by TiTi** [comment]
> [@TiTi](https://github.com/TiTi) reading [your comment above](https://github.com/microsoft/azure-pipelines-tasks/issues/15473#issuecomment-1970789538), do I read it correctly that you managed to force color output (even though not properly escaped) with azure cli tools (i.e. `az deployment sub create ... --what-if`)?
> 
> Was wondering if colors could be forced with azure cli based deployment, but didn't get so far yet...

i did not manage to get color output in azure devops
if someone knows, i'm all ears

**Comment by 0-0-1-0-1-0-1-0** [comment]
Yes if you look at my post everything you need to get color output is there you need to use the powershell modules rather than the Azure cli to do the whatif analysis.I think I covered it pretty well, but you need to use the write-color module i included in my post. I put this in my self hosted servers under the default path so it'd always loaded with all modules when powershell is ran on the build agents.Then use the write-color module as before but simply add -ANSI parameter to your scripts (i do in line scripts so this is super easy).Regardless if you use the code I pasted above with changes for your environment. Color output works proper.In fact you can get color in any powershell script in the Azure Pipelines log console I'll see if I can provide a screenshot as proof.Thank you,Greg Hypta
-------- Original message --------From: Thibault ROHMER ***@***.***> Date: 7/19/25  2:48 PM  (GMT-06:00) To: microsoft/azure-pipelines-tasks ***@***.***> Cc: 0-0-1-0-1-0-1-0 ***@***.***>, Comment ***@***.***> Subject: Re: [microsoft/azure-pipelines-tasks] Azure Pipelines Azure CLI task - no colors when doing Bicep deployment (Issue #15473) TiTi left a comment (microsoft/azure-pipelines-tasks#15473)

@TiTi reading your comment above, do I read it correctly that you managed to force color output (even though not properly escaped) with azure cli tools (i.e. az deployment sub create ... --what-if)?
Was wondering if colors could be forced with azure cli based deployment, but didn't get so far yet...

i did not manage to get color output in azure devops
if someone knows, i'm all ears

**Comment by 0-0-1-0-1-0-1-0** [comment]
> [@TiTi](https://github.com/TiTi) reading [your comment above](https://github.com/microsoft/azure-pipelines-tasks/issues/15473#issuecomment-1970789538), do I read it correctly that you managed to force color output (even though not properly escaped) with azure cli tools (i.e. `az deployment sub create ... --what-if`)?
> 
> Was wondering if colors could be forced with azure cli based deployment, but didn't get so far yet...

> > [@TiTi](https://github.com/TiTi) reading [your comment above](https://github.com/microsoft/azure-pipelines-tasks/issues/15473#issuecomment-1970789538), do I read it correctly that you managed to force color output (even though not properly escaped) with azure cli tools (i.e. `az deployment sub create ... --what-if`)?
> > Was wondering if colors could be forced with azure cli based deployment, but didn't get so far yet...
> 
> i did not manage to get color output in azure devops if someone knows, i'm all ears

See my post on how to get it to work properly:

https://github.com/microsoft/azure-pipelines-tasks/issues/15473#issuecomment-2749306594

If you want to support color in your powrshell scripts so that your Azure Pipelines console logs actually have color you can reference this post:

https://github.com/microsoft/azure-pipelines-tasks/issues/15473#issuecomment-2749593724

**Comment by tknerr** [comment]
Thanks @0-0-1-0-1-0-1-0  👍

I saw your comment above, but was wondering if someone managed to also get colored ouput with plain azure cli commandline tools in Azure DevOps (i.e. using `az deployment sub create ...` instead of the corresponding powershell Cmdlets).

**Comment by TiTi** [comment]
Ok so i took some time to **try** to switch my pipeline from `az cli` deployment to Az powershell.  
I had a lot of issues to set up everything's needed in our private linux agents, that why i did not do the switch until now btw. Definitely painful for multiple reasons (binary to dl, path or alias, linux agent, az module vs sysprep/deprovision, ...). Anyhow managed to.

The `pwsh -Command {}` is really tricky and fails with weird errors as soon as you put 2 things, or use the script syntax.
Like:
`$output = pwsh -Command { Set-AzContext -Subscription "${{ parameters.subscription }}" | Out-Null; New-AzResourceGroupDeployment -Name "${{ parameters.deploymentName }}" -ResourceGroupName "${{ parameters.resourceGroupName }}" -TemplateParameterFile "${{ parameters.bicepParam }}" -WhatIf -WhatIfExcludeChangeType Ignore}`
Got:
> Cannot retrieve the dynamic parameters for the cmdlet. Run Connect-AzAccount to login.

Which is super confusing as it works without the extra pwsh context, directly in inline script of AzurePowerShell@5

As i have multiple subscriptions, I first need to do: `Set-AzContext -Subscription "${{ parameters.subscription }}" | Out-Null` to switch to the proper one!

Same errors with the script syntax:
```
        $script = @"
        Set-AzContext -Subscription "${{ parameters.subscription }}" | Out-Null
        New-AzResourceGroupDeployment -Name "${{ parameters.deploymentName }}" -ResourceGroupName "${{ parameters.resourceGroupName }}" -TemplateParameterFile "${{ parameters.bicepParam }}" -WhatIf -WhatIfExcludeChangeType Ignore
        "@
        $output = $script | pwsh -Command -
```

So the only working way was:
```
        Set-AzContext -Subscription "${{ parameters.subscription }}" | Out-Null
        $output = pwsh -Command {New-AzResourceGroupDeployment -Name "${{ parameters.deploymentName }}" -ResourceGroupName "${{ parameters.resourceGroupName }}" -TemplateParameterFile "${{ parameters.bicepParam }}" -WhatIf -WhatIfExcludeChangeType Ignore}
```
which is in the correct subscription

---

And then yeah I get the color but it's a bit different than my terminal

my terminal:
<img width="1257" height="426" alt="Image" src="https://github.com/user-attachments/assets/fed12a1b-b70a-4a1b-9a44-dbd082096de4" />

in azure devops:
<img width="1465" height="401" alt="Image" src="https://github.com/user-attachments/assets/50625ec2-f1ee-4f13-805f-520dad77e35e" />

@0-0-1-0-1-0-1-0  any idea why the colored output differs?? (it's not green for the properties name)
(i used your initial script with multiple `-replace`)

Anyway, thanks now i have some kind of color :)

**Comment by 0-0-1-0-1-0-1-0** [comment]
> Ok so i took some time to **try** to switch my pipeline from `az cli` deployment to Az powershell. I had a lot of issues to set up everything's needed in our private linux agents, that why i did not do the switch until now btw. Definitely painful for multiple reasons (binary to dl, path or alias, linux agent, az module vs sysprep/deprovision, ...). Anyhow managed to.
> 
> The `pwsh -Command {}` is really tricky and fails with weird errors as soon as you put 2 things, or use the script syntax. Like: `$output = pwsh -Command { Set-AzContext -Subscription "${{ parameters.subscription }}" | Out-Null; New-AzResourceGroupDeployment -Name "${{ parameters.deploymentName }}" -ResourceGroupName "${{ parameters.resourceGroupName }}" -TemplateParameterFile "${{ parameters.bicepParam }}" -WhatIf -WhatIfExcludeChangeType Ignore}` Got:
> 
> > Cannot retrieve the dynamic parameters for the cmdlet. Run Connect-AzAccount to login.
> 
> Which is super confusing as it works without the extra pwsh context, directly in inline script of AzurePowerShell@5
> 
> As i have multiple subscriptions, I first need to do: `Set-AzContext -Subscription "${{ parameters.subscription }}" | Out-Null` to switch to the proper one!
> 
> Same errors with the script syntax:
> 
> ```
>         $script = @"
>         Set-AzContext -Subscription "${{ parameters.subscription }}" | Out-Null
>         New-AzResourceGroupDeployment -Name "${{ parameters.deploymentName }}" -ResourceGroupName "${{ parameters.resourceGroupName }}" -TemplateParameterFile "${{ parameters.bicepParam }}" -WhatIf -WhatIfExcludeChangeType Ignore
>         "@
>         $output = $script | pwsh -Command -
> ```
> 
> So the only working way was:
> 
> ```
>         Set-AzContext -Subscription "${{ parameters.subscription }}" | Out-Null
>         $output = pwsh -Command {New-AzResourceGroupDeployment -Name "${{ parameters.deploymentName }}" -ResourceGroupName "${{ parameters.resourceGroupName }}" -TemplateParameterFile "${{ parameters.bicepParam }}" -WhatIf -WhatIfExcludeChangeType Ignore}
> ```
> 
> which is in the correct subscription
> 
> And then yeah I get the color but it's a bit different than my terminal
> 
> my terminal: <img alt="Image" width="1257" height="426" src="https://private-user-images.githubusercontent.com/137691/469855603-fed12a1b-b70a-4a1b-9a44-dbd082096de4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMzMDM3MDksIm5iZiI6MTc1MzMwMzQwOSwicGF0aCI6Ii8xMzc2OTEvNDY5ODU1NjAzLWZlZDEyYTFiLWI3MGEtNGExYi05YTQ0LWRiZDA4MjA5NmRlNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyM1QyMDQzMjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hOTA4N2ZkZjhiZmY1NTkxNzQ4MWE1ODJhYzYxNzJiOTJiNDcyYjc1N2U3YzM1N2E5NzE5OGNmODljNTY5YjI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.GALLXaUrlwZgXK4bcpWbrJhEQZIVPOcSpF0cGBydkz8">
> 
> in azure devops: <img alt="Image" width="1465" height="401" src="https://private-user-images.githubusercontent.com/137691/469856563-50625ec2-f1ee-4f13-805f-520dad77e35e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMzMDM3MDksIm5iZiI6MTc1MzMwMzQwOSwicGF0aCI6Ii8xMzc2OTEvNDY5ODU2NTYzLTUwNjI1ZWMyLWYxZWUtNGYxMy04MDVmLTUyMGRhZDc3ZTM1ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyM1QyMDQzMjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04YWU2YzJhNzIwNDY5M2E0YjI2ODAyMWM3ZjNlODNjNDEyOWMzMzAzNmQ3MGZkNDJmNmI1Y2FlYjg2ZWI3MGJiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4fztdVFWqq1uEp8C2i257BZAJemWkR18C1cRkSv0qyU">
> 
> [@0-0-1-0-1-0-1-0](https://github.com/0-0-1-0-1-0-1-0) any idea why the colored output differs?? (it's not green for the properties name) (i used your initial script with multiple `-replace`)
> 
> Anyway, thanks now i have some kind of color :)

So the reason it doesn't work because the pwsh -command () is actually starting another powershell process, this process has no idea about any other information, hence the ask to az login if you want to switch subscriptions. It only knows about the subscription that the original powershell process had. What you did would be the only afaik to make this work.

As for the coloring that's how it comes out on windows from what i remember, and I think its easier to read left side is the parameter, right side is the value that is changing. Especially when you have modifications and not just adds.

I could try to make the whole block green like your first screenshot, but it purely depends on the output logs if that is possible at all.

Glad it works. 😃

**Comment by 0-0-1-0-1-0-1-0** [comment]
> Thanks [@0-0-1-0-1-0-1-0](https://github.com/0-0-1-0-1-0-1-0) 👍
> 
> I saw your comment above, but was wondering if someone managed to also get colored ouput with plain azure cli commandline tools in Azure DevOps (i.e. using `az deployment sub create ...` instead of the corresponding powershell Cmdlets).

Sadly I dont think there is a way since the az cli output seems to be different than the powershell module.

I can give it a try again but I think the format of the output is different between the two. So far only way I got this to work is how I described it above.

**Comment by ingenobajon** [comment]
Please, do not close this issue.

**Comment by slavizh** [comment]
still needed.

**Comment by anthony-c-martin** [comment]
Has anyone tried using the [`BicepDeploy@0` ADO task](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/bicep-deploy-v0?view=azure-pipelines) instead, and can confirm whether this shows colors properly?

I know it's not a solution to this specific issue, but may be a workaround for some people.
